src/main/java/uk/gov/ons/ctp/response/action/service/ActionService.java

```diff
@@ -173,25 +200,129 @@ public Action createAction(final Action action) {
   }
 
   @Transactional
-  public void createScheduledActions(Integer actionPlanJobId) {
-    final ActionPlanJob actionPlanJob =
-        actionPlanJobRepository.findByActionPlanJobPK(actionPlanJobId);
+  public void createScheduledActions(ActionPlan actionPlan, ActionPlanJob actionPlanJob) {
+    List<ActionCase> cases = actionCaseRepo.findByActionPlanFK(actionPlan.getActionPlanPK());
+    List<ActionRule> rules = actionRuleRepo.findByActionPlanFK(actionPlan.getActionPlanPK());
+
+    // For each case/rule pair create actions
+    cases.forEach(
+        caze -> {
+          if (isActionPlanLive(caze)) {
+            rules.forEach(
+                rule -> {
+                  if (hasRuleTriggered(rule)) {
+                    try {
+                      createActions(caze, rule);
+                    } catch (Exception ex) {
+                      log.with("caseId", caze.getId().toString())
+                          .with("actionRuleId", rule.getId().toString())
+                          .error(ex.toString());
+                    }
+                  }
+                });
+          }
+        });
+    updatePlanAndJob(actionPlan, actionPlanJob);
+  }
 
-    if (actionPlanJob == null) {
-      return;
-    }
+  private boolean isActionPlanLive(ActionCase actionCase) {
+    final Timestamp currentTime = new Timestamp((new Date()).getTime());
+    return actionCase.getActionPlanStartDate().before(currentTime)
+        && actionCase.getActionPlanEndDate().after(currentTime);
+  }
 
-    final ActionPlan actionPlan =
-        actionPlanRepository.findByActionPlanPK(actionPlanJob.getActionPlanFK());
+  private boolean hasRuleTriggered(ActionRule rule) {
     final Timestamp currentTime = new Timestamp((new Date()).getTime());
+    Timestamp triggerDateTime =
+        Timestamp.valueOf(
+            LocalDateTime.ofInstant(rule.getTriggerDateTime().toInstant(), ZoneOffset.UTC));
+    return triggerDateTime.before(currentTime);
+  }
+
+  private void createActions(ActionCase actionCase, ActionRule actionRule) {
+    ActionType actionType = actionTypeRepo.findByActionTypePK(actionRule.getActionTypeFK());
+
+    // If creating actions for business respondents, create an action per respondent
+    // Else create single action
+    if (actionCase.getSampleUnitType().equals(SampleUnitDTO.SampleUnitType.B.toString())
+        && (actionType.getActionTypePK() == 5 || actionType.getActionTypePK() == 7)) {
```

Comment:
Magic numbers - should be constants

Reply:
In fact, can we look these up in a more sophisticated way, in case the PK changes?

Reply:
:+1: You're right, here we should be checking on the Type Handler rather than typePK